src/D2L.CodeStyle.Analyzers/Diagnostics.cs

```diff
@@ -514,5 +514,77 @@ public static class Diagnostics {
 			isEnabledByDefault: true,
 			description: "We require that event types be sealed."
 		);
+
+		public static readonly DiagnosticDescriptor MutableBaseType = new DiagnosticDescriptor(
+			id: "D2L0061",
+			title: "Base class for immutable type must be [Immutable] (or [ImmutableBaseClass])",
+			messageFormat: "{0}'s base class is {1} which is missing [Immutable] (or, more weakly, [ImmutableBaseClass]).",
+			category: "Immutability",
+			defaultSeverity: DiagnosticSeverity.Error,
+			isEnabledByDefault: true
+		);
+
+		public static readonly DiagnosticDescriptor ArraysAreMutable = new DiagnosticDescriptor(
+			id: "D2L0062",
+			title: "Arrays are mutable and thus can't be held by an immutable type.",
+			messageFormat: "{0}[]'s are mutable even if {0}'s are immutable because array items can be substituted. Prefer an immutable container type such as ImmutableArray<{0}>.",
+			category: "Immutability",
+			defaultSeverity: DiagnosticSeverity.Error,
+			isEnabledByDefault: true
+		);
+
+		public static readonly DiagnosticDescriptor DynamicObjectsAreMutable = new DiagnosticDescriptor(
+			id: "D2L0063",
+			title: "Dynamic objects are mutable and thus can't be held by an immutable type.",
+			messageFormat: "Dynamic objects are mutable and thus can't be held by an immutable type.",
+			category: "Immutability",
+			defaultSeverity: DiagnosticSeverity.Error,
+			isEnabledByDefault: true
+		);
+
+		public static readonly DiagnosticDescriptor TypeParameterIsNotKnownToBeImmutable = new DiagnosticDescriptor(
+			id: "D2L0064",
+			title: "Type parameter is not known to be immutable.",
+			messageFormat: "The type parameter {0} is not known to be immutable. Add [Immutable] to it.",
+			category: "Immutability",
+			defaultSeverity: DiagnosticSeverity.Error,
+			isEnabledByDefault: true
+		);
+
+		public static readonly DiagnosticDescriptor UnexpectedTypeKind = new DiagnosticDescriptor(
+			id: "D2L0065",
+			title: "Unsupported type kind inside an immutable type declaration.",
+			messageFormat: "The type kind {0} is unexpected/unhandled by the immutability analyzer, and for safety we are emitting an error. This is a bug in the analyzer.",
+			category: "Immutability",
+			defaultSeverity: DiagnosticSeverity.Error,
+			isEnabledByDefault: true
+		);
+
+		public static readonly DiagnosticDescriptor TypeNeedsImmutableAttribute = new DiagnosticDescriptor(
```

Comment:
The naming of this implies to me it would be ~ D2L0040 MissingTransitiveImmutableAttribute vs what it is

Reply:
Ah, these are two different cases:

1. If you need `[Immutable]` because you implement an `[Immutable]` interface (or etc.) then you should get D2L0040
2. If you are held onto a `Foo` and `Foo` isn't immutable, that's an error

Ideally we only show (1) and not (2) if applicable.

If you can think of a better name for the diagnostic var LMK.

Reply:
Yeah for sure. Maybe `MutableTypeHeldByImmutableType`

Reply:
Done